Participate in iminuit v2.0 beta?

Dear pyhf team,

I am about to finish a major rewrite of iminuit, version 2.0, that replaces Cython as the tool to wrap C++ Minuit2 with pybind11, which is going to solve several issues that the legacy code had. All the good things that this will bring are listed on top of this PR:
scikit-hep/iminuit#502

Switching to the new version of iminuit should be completely transparent to you, since the new version passes the comprehensive suite of unit tests of iminuit-v1.x. However, I would like to use this opportunity to finally remove interface that has been successively marked as deprecated in versions 1.3 to 1.5.

Therefore my two question to you:

 * Did you take note of the deprecation warnings in iminuit and did you keep up with the interface changes so far?
 * Are you interested in trying out a Beta release of v2.0 to work out any possible bugs in the new version before the release?

Best regards,
Hans, iminuit maintainer

> I merged [scikit-hep/iminuit#514](https://github.com/scikit-hep/iminuit/pull/514), which fixes the bug, but you still need to change the counter names as described in my previous message.

Okay, with that we're now passing that section but failing 

https://github.com/scikit-hep/pyhf/blob/28fdfe95a3a4846ba70a9a338b3f72a94eac1322/tests/test_optim.py#L47

with

```pytb
>           result = pyhf.infer.mle.fit(data, m, do_grad=do_grad, do_stitch=do_stitch)

tests/test_optim.py:105: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/pyhf/infer/mle.py:122: in fit
    return opt.minimize(
src/pyhf/optimize/mixins.py:136: in minimize
    result = self._internal_minimize(**minimizer_kwargs, options=kwargs)
src/pyhf/optimize/mixins.py:38: in _internal_minimize
    result = self._minimize(
src/pyhf/optimize/opt_minuit.py:101: in _minimize
    minimizer.migrad(ncall=maxiter)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <iminuit._minuit.Minuit object at 0x7f57644a9f40>, ncall = 100000
precision = None, iterate = 5

    def migrad(self, ncall=None, precision=None, iterate=5):
        """Run MIGRAD.
    
        MIGRAD is a robust minimisation algorithm which earned its reputation
        in 40+ years of almost exclusive usage in high-energy physics. How
        MIGRAD works is described in the `MINUIT paper`_.
    
        **Arguments:**
    
            * **ncall**: integer or None, optional; (approximate)
              maximum number of call before MIGRAD will stop trying. Default: None
              (indicates to use MIGRAD's internal heuristic). Note: MIGRAD may slightly
              violate this limit, because it checks the condition only after a full
              iteration of the algorithm, which usually performs several function calls.
    
            * **precision**: override Minuit precision estimate for the cost function.
              Default: None (= use epsilon of a C++ double). If the cost function has a
              lower precision (e.g. of a C++ float), setting this to a lower value will
              accelerate convergence and reduce the rate of unsuccessful convergence.
    
            * **iterate**: automatically call Migrad up to N times if convergence
              was not reached. Default: 5. This simple heuristic makes Migrad converge
              more often even if the numerical precision of the cost function is low.
              Setting this to 1 disables the feature.
    
        **Return:**
    
            :ref:`function-minimum-sruct`, list of :ref:`minuit-param-struct`
        """
        if ncall is None:
            ncall = 0  # tells C++ Minuit to use its internal heuristic
    
        if iterate < 1:
            raise ValueError("iterate must be at least 1")
    
        migrad = MnMigrad(self._fcn, self._last_state, self.strategy)
        migrad.set_print_level(self.print_level)
        if precision is not None:
            migrad.precision = precision
    
        # Automatically call Migrad up to `iterate` times if minimum is not valid.
        # This simple heuristic makes Migrad converge more often.
        for _ in range(iterate):
>           fm = migrad(ncall, self.tol)
E           RuntimeError: Unable to cast Python instance to C++ type (compile in debug mode for details)
```

@HDembinski I'm now also seeing in runs

```pytb
    def test_hypotest_backends(backend, kwargs):
        """
        Check that hypotest runs fully across all backends for all calculator types.
        """
        pdf = pyhf.simplemodels.hepdata_like(
            signal_data=[12.0, 11.0], bkg_data=[50.0, 52.0], bkg_uncerts=[3.0, 7.0]
        )
        data = [51, 48] + pdf.config.auxdata
>       assert pyhf.infer.hypotest(1.0, data, pdf, **kwargs) is not None

tests/test_infer.py:165: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/pyhf/infer/__init__.py:146: in hypotest
    teststat = calc.teststatistic(poi_test)
src/pyhf/infer/calculators.py:258: in teststatistic
    qmu_v = teststat_func(
src/pyhf/infer/test_statistics.py:177: in qmu_tilde
    return _qmu_like(mu, data, pdf, init_pars, par_bounds, fixed_params)
src/pyhf/infer/test_statistics.py:19: in _qmu_like
    tmu_like_stat, (_, muhatbhat) = _tmu_like(
src/pyhf/infer/test_statistics.py:38: in _tmu_like
    mubhathat, fixed_poi_fit_lhood_val = fixed_poi_fit(
src/pyhf/infer/mle.py:190: in fixed_poi_fit
    return fit(data, pdf, init_pars, par_bounds, fixed_params, **kwargs)
src/pyhf/infer/mle.py:122: in fit
    return opt.minimize(
src/pyhf/optimize/mixins.py:136: in minimize
    result = self._internal_minimize(**minimizer_kwargs, options=kwargs)
src/pyhf/optimize/mixins.py:35: in _internal_minimize
    minimizer = self._get_minimizer(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pyhf.optimize.minuit_optimizer object at 0x7f3fa6e957c0>
objective_and_grad = <function wrap_objective.<locals>.func at 0x7f3f84139b80>
init_pars = [1.0, 1.0, 1.0]
init_bounds = [(0, 10), (1e-10, 10.0), (1e-10, 10.0)], fixed_vals = [(0, 1.0)]
do_grad = False

    def _get_minimizer(
        self, objective_and_grad, init_pars, init_bounds, fixed_vals=None, do_grad=False
    ):
    
        step_sizes = [(b[1] - b[0]) / float(self.steps) for b in init_bounds]
        fixed_vals = fixed_vals or []
        # Minuit wants True/False for each parameter
        fixed_bools = [False] * len(init_pars)
        for index, val in fixed_vals:
            fixed_bools[index] = True
            init_pars[index] = val
            step_sizes[index] = 0.0
    
        # Minuit requires jac=callable
        if do_grad:
            wrapped_objective = lambda pars: objective_and_grad(pars)[0]
            jac = lambda pars: objective_and_grad(pars)[1]
        else:
            wrapped_objective = objective_and_grad
            jac = None
    
        kwargs = dict(
            fcn=wrapped_objective,
            grad=jac,
            start=init_pars,
            error=step_sizes,
            limit=init_bounds,
            fix=fixed_bools,
            print_level=self.verbose,
            errordef=self.errordef,
        )
>       return iminuit.Minuit.from_array_func(**kwargs)
E       AttributeError: type object 'Minuit' has no attribute 'from_array_func'
```

It looks like https://github.com/scikit-hep/iminuit/pull/526/ removed `from_array_func`, so is there a new API that will allow for similar behavior to `iminuit.Minuit.from_array_func(**kwargs)`?

Cross-posting with the PR:

```pytb
_______________ test_optimizer_stitching[minuit-pytorch_backend] _______________

backend = <class 'pyhf.tensor.pytorch_backend.pytorch_backend'>
optimizer = 'minuit'

    @pytest.mark.parametrize(
        'backend',
        [
            pyhf.tensor.numpy_backend,
            pyhf.tensor.jax_backend,
            pyhf.tensor.tensorflow_backend,
            pyhf.tensor.pytorch_backend,
        ],
    )
    @pytest.mark.parametrize('optimizer', ['scipy', 'minuit'])
    def test_optimizer_stitching(backend, optimizer):
        pyhf.set_backend(backend(precision='64b'), optimizer)
    
        pdf = pyhf.simplemodels.hepdata_like([50.0], [100.0], [10])
        data = [125.0] + pdf.config.auxdata
    
>       result_nostitch = pyhf.infer.mle.fixed_poi_fit(2.0, data, pdf, do_stitch=False)

tests/test_validation.py:848: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/pyhf/infer/mle.py:190: in fixed_poi_fit
    return fit(data, pdf, init_pars, par_bounds, fixed_params, **kwargs)
src/pyhf/infer/mle.py:122: in fit
    return opt.minimize(
src/pyhf/optimize/mixins.py:136: in minimize
    result = self._internal_minimize(**minimizer_kwargs, options=kwargs)
src/pyhf/optimize/mixins.py:38: in _internal_minimize
    result = self._minimize(
src/pyhf/optimize/opt_minuit.py:129: in _minimize
    minimizer.migrad(ncall=maxiter)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = (Param(number=0, name='x0', value=2.0, error=0.0, merror=None, is_const=False, is_fixed=True, has_limits=True, has_low...lse, is_fixed=False, has_limits=True, has_lower_limit=True, has_upper_limit=True, lower_limit=1e-10, upper_limit=10.0))
ncall = 100000, iterate = 5

    def migrad(self, ncall=None, iterate=5):
        """Run MnMigrad from the Minuit2 library.
    
        MIGRAD is a robust minimisation algorithm which earned its reputation
        in 40+ years of almost exclusive usage in high-energy physics. How
        MIGRAD works is described in the `MINUIT paper`_.
    
        **Arguments:**
    
            * **ncall**: integer or None, optional; (approximate)
              maximum number of calls before minimization will be aborted. Default: None
              (indicates to use an internal heuristic). Note: The limit may be slightly
              violated, because the condition is checked only after a full iteration of
              the algorithm, which usually performs several function calls.
    
            * **iterate**: automatically call Migrad up to N times if convergence
              was not reached. Default: 5. This simple heuristic makes Migrad converge
              more often even if the numerical precision of the cost function is low.
              Setting this to 1 disables the feature.
    
        **Return:**
    
            self
        """
        if ncall is None:
            ncall = 0  # tells C++ Minuit to use its internal heuristic
    
        if iterate < 1:
            raise ValueError("iterate must be at least 1")
    
        migrad = MnMigrad(self._fcn, self._last_state, self.strategy)
    
        # Automatically call Migrad up to `iterate` times if minimum is not valid.
        # This simple heuristic makes Migrad converge more often.
        for _ in range(iterate):
            # workaround: precision must be set again after each call to MnMigrad
            if self._precision is not None:
                migrad.precision = self._precision
>           fm = migrad(ncall, self._tolerance)
E           RuntimeError: Unable to cast Python instance to C++ type (compile in debug mode for details)

/opt/hostedtoolcache/Python/3.8.6/x64/lib/python3.8/site-packages/iminuit/_minuit.py:547: RuntimeError
```

--- a/setup.py
+++ b/setup.py
@@ -9,7 +9,7 @@ extras_require = {
     'torch': ['torch~=1.2'],
     'jax': ['jax~=0.2.4', 'jaxlib~=0.1.56'],
     'xmlio': ['uproot3~=3.14'],  # Future proof against uproot4 API changes
-    'minuit': ['iminuit~=2.0'],
+    'minuit': ['iminuit~=2.1'],
 }
 extras_require['backends'] = sorted(
     set(

--- a/setup.py
+++ b/setup.py
@@ -9,7 +9,7 @@ extras_require = {
     'torch': ['torch~=1.2'],
     'jax': ['jax~=0.2.4', 'jaxlib~=0.1.56'],
     'xmlio': ['uproot3~=3.14'],  # Future proof against uproot4 API changes
-    'minuit': ['iminuit~=1.5.3'],
+    'minuit': ['iminuit~=2.0'],
 }
 extras_require['backends'] = sorted(
     set(

--- a/src/pyhf/optimize/opt_minuit.py
+++ b/src/pyhf/optimize/opt_minuit.py
@@ -141,7 +141,7 @@ class minuit_optimizer(OptimizerMixin):
             fun=minimizer.fval,
             hess_inv=hess_inv,
             message=message,
-            nfev=minimizer.ncalls_total,
-            njev=minimizer.ngrads_total,
+            nfev=minimizer.nfcn,
+            njev=minimizer.ngrads,
             minuit=minimizer,
         )

--- a/src/pyhf/optimize/opt_minuit.py
+++ b/src/pyhf/optimize/opt_minuit.py
@@ -58,9 +58,6 @@ class minuit_optimizer(OptimizerMixin):
             wrapped_objective = objective_and_grad
             jac = None
 
-        # FAILING:
-        # * tests/test_optim.py
-        # * tests/test_validation.py
         minuit = iminuit.Minuit(wrapped_objective, init_pars, grad=jac)
         minuit.errors = step_sizes
         minuit.limits = init_bounds
@@ -112,7 +109,7 @@ class minuit_optimizer(OptimizerMixin):
         minimizer.tol = tolerance
         minimizer.migrad(ncall=maxiter)
         # Following lines below come from:
-        # https://github.com/scikit-hep/iminuit/blob/f12cd519aa08e244ad71f56fe7a97d42117196e4/src/iminuit/_minimize.py#L111-L130
+        # https://github.com/scikit-hep/iminuit/blob/23bad7697e39d363f259ca8349684df939b1b2e6/src/iminuit/_minimize.py#L111-L130
         message = "Optimization terminated successfully."
         if not minimizer.valid:
             message = "Optimization failed."

--- a/src/pyhf/optimize/opt_minuit.py
+++ b/src/pyhf/optimize/opt_minuit.py
@@ -143,14 +143,14 @@ class minuit_optimizer(OptimizerMixin):
         if minimizer.valid:
             # Extra call to hesse() after migrad() is always needed for good error estimates. If you pass a user-provided gradient to MINUIT, convergence is faster.
             minimizer.hesse()
-            hess_inv = minimizer.covariance()
+            hess_inv = minimizer.covariance
 
         unc = None
         if return_uncertainties:
             unc = minimizer.np_errors()
 
         return scipy.optimize.OptimizeResult(
-            x=minimizer.np_values(),
+            x=minimizer.values,
             unc=unc,
             success=minimizer.valid,
             fun=minimizer.fval,

--- a/src/pyhf/optimize/opt_minuit.py
+++ b/src/pyhf/optimize/opt_minuit.py
@@ -58,6 +58,9 @@ class minuit_optimizer(OptimizerMixin):
             wrapped_objective = objective_and_grad
             jac = None
 
+        # FAILING:
+        # * tests/test_optim.py
+        # * tests/test_validation.py
         minuit = iminuit.Minuit(wrapped_objective, init_pars, grad=jac)
         minuit.errors = step_sizes
         minuit.limits = init_bounds

--- a/src/pyhf/optimize/opt_minuit.py
+++ b/src/pyhf/optimize/opt_minuit.py
@@ -58,23 +58,24 @@ class minuit_optimizer(OptimizerMixin):
             wrapped_objective = objective_and_grad
             jac = None
 
-        # kwargs = dict(
-        #     fcn=wrapped_objective,
-        #     grad=jac,
-        #     start=init_pars,
-        #     error=step_sizes,
-        #     limit=init_bounds,
-        #     fix=fixed_bools,
-        #     print_level=self.verbose,
-        #     errordef=self.errordef,
-        # )
-        minuit = iminuit.Minuit(fcn=wrapped_objective, grad=jac)
-        minuit.start = init_pars
-        minuit.error = step_sizes
-        minuit.limit = init_bounds
-        minuit.fix = fixed_bools
-        minuit.print_level = self.verbose
-        minuit.errordef = self.errordef
+        kwargs = dict(
+            fcn=wrapped_objective,
+            grad=jac,
+            start=init_pars,
+            error=step_sizes,
+            limit=init_bounds,
+            fix=fixed_bools,
+            print_level=self.verbose,
+            errordef=self.errordef,
+        )
+        minuit = iminuit.Minuit(**kwargs)
+        # minuit = iminuit.Minuit(fcn=wrapped_objective, grad=jac)
+        # minuit.start = init_pars
+        # minuit.error = step_sizes
+        # minuit.limit = init_bounds
+        # minuit.fix = fixed_bools
+        # minuit.print_level = self.verbose
+        # minuit.errordef = self.errordef
         return minuit
 
     def _minimize(

--- a/src/pyhf/optimize/opt_minuit.py
+++ b/src/pyhf/optimize/opt_minuit.py
@@ -68,21 +68,21 @@ class minuit_optimizer(OptimizerMixin):
         #     print_level=self.verbose,
         #     errordef=self.errordef,
         # )
-        kwargs = dict(
-            error=step_sizes,
-            limit=init_bounds,
-            fix=fixed_bools,
-            print_level=self.verbose,
-            errordef=self.errordef,
-        )
+        # kwargs = dict(
+        #     error=step_sizes,
+        #     limit=init_bounds,
+        #     fix=fixed_bools,
+        #     print_level=self.verbose,
+        #     errordef=self.errordef,
+        # )
         # minuit = iminuit.Minuit(**kwargs)
-        minuit = iminuit.Minuit(wrapped_objective, init_pars, grad=jac, **kwargs)
+        minuit = iminuit.Minuit(wrapped_objective, init_pars, grad=jac)
         # minuit.start = init_pars
-        # minuit.error = step_sizes
-        # minuit.limit = init_bounds
-        # minuit.fix = fixed_bools
-        # minuit.print_level = self.verbose
-        # minuit.errordef = self.errordef
+        minuit.errors = step_sizes
+        minuit.limits = init_bounds
+        minuit.fixed = fixed_bools
+        minuit.print_level = self.verbose
+        minuit.errordef = self.errordef
         return minuit
 
     def _minimize(

--- a/src/pyhf/optimize/opt_minuit.py
+++ b/src/pyhf/optimize/opt_minuit.py
@@ -142,6 +142,6 @@ class minuit_optimizer(OptimizerMixin):
             hess_inv=hess_inv,
             message=message,
             nfev=minimizer.nfcn,
-            njev=minimizer.ngrads,
+            njev=minimizer.ngrad,
             minuit=minimizer,
         )

--- a/src/pyhf/optimize/opt_minuit.py
+++ b/src/pyhf/optimize/opt_minuit.py
@@ -58,18 +58,25 @@ class minuit_optimizer(OptimizerMixin):
             wrapped_objective = objective_and_grad
             jac = None
 
+        # kwargs = dict(
+        #     fcn=wrapped_objective,
+        #     grad=jac,
+        #     start=init_pars,
+        #     error=step_sizes,
+        #     limit=init_bounds,
+        #     fix=fixed_bools,
+        #     print_level=self.verbose,
+        #     errordef=self.errordef,
+        # )
         kwargs = dict(
-            fcn=wrapped_objective,
-            grad=jac,
-            start=init_pars,
             error=step_sizes,
             limit=init_bounds,
             fix=fixed_bools,
             print_level=self.verbose,
             errordef=self.errordef,
         )
-        minuit = iminuit.Minuit(**kwargs)
-        # minuit = iminuit.Minuit(fcn=wrapped_objective, grad=jac)
+        # minuit = iminuit.Minuit(**kwargs)
+        minuit = iminuit.Minuit(wrapped_objective, init_pars, grad=jac, **kwargs)
         # minuit.start = init_pars
         # minuit.error = step_sizes
         # minuit.limit = init_bounds

--- a/src/pyhf/optimize/opt_minuit.py
+++ b/src/pyhf/optimize/opt_minuit.py
@@ -141,7 +141,7 @@ class minuit_optimizer(OptimizerMixin):
             fun=minimizer.fval,
             hess_inv=hess_inv,
             message=message,
-            nfev=minimizer.ncalls_total,
-            njev=minimizer.ngrads_total,
+            nfev=minimizer.nfcn,
+            njev=minimizer.ngrad,
             minuit=minimizer,
         )

--- a/src/pyhf/optimize/opt_minuit.py
+++ b/src/pyhf/optimize/opt_minuit.py
@@ -58,26 +58,7 @@ class minuit_optimizer(OptimizerMixin):
             wrapped_objective = objective_and_grad
             jac = None
 
-        # kwargs = dict(
-        #     fcn=wrapped_objective,
-        #     grad=jac,
-        #     start=init_pars,
-        #     error=step_sizes,
-        #     limit=init_bounds,
-        #     fix=fixed_bools,
-        #     print_level=self.verbose,
-        #     errordef=self.errordef,
-        # )
-        # kwargs = dict(
-        #     error=step_sizes,
-        #     limit=init_bounds,
-        #     fix=fixed_bools,
-        #     print_level=self.verbose,
-        #     errordef=self.errordef,
-        # )
-        # minuit = iminuit.Minuit(**kwargs)
         minuit = iminuit.Minuit(wrapped_objective, init_pars, grad=jac)
-        # minuit.start = init_pars
         minuit.errors = step_sizes
         minuit.limits = init_bounds
         minuit.fixed = fixed_bools

--- a/src/pyhf/optimize/opt_minuit.py
+++ b/src/pyhf/optimize/opt_minuit.py
@@ -113,7 +113,7 @@ class minuit_optimizer(OptimizerMixin):
         minimizer.tol = tolerance
         minimizer.migrad(ncall=maxiter)
         # Following lines below come from:
-        # https://github.com/scikit-hep/iminuit/blob/64acac11cfa2fb91ccbd02d1b3c51f8a9e2cc484/src/iminuit/_minimize.py#L102-L121
+        # https://github.com/scikit-hep/iminuit/blob/f12cd519aa08e244ad71f56fe7a97d42117196e4/src/iminuit/_minimize.py#L111-L130
         message = "Optimization terminated successfully."
         if not minimizer.valid:
             message = "Optimization failed."
@@ -141,7 +141,7 @@ class minuit_optimizer(OptimizerMixin):
             fun=minimizer.fval,
             hess_inv=hess_inv,
             message=message,
-            nfev=minimizer.nfcn,
-            njev=minimizer.ngrad,
+            nfev=minimizer.ncalls_total,
+            njev=minimizer.ngrads_total,
             minuit=minimizer,
         )

--- a/src/pyhf/optimize/opt_minuit.py
+++ b/src/pyhf/optimize/opt_minuit.py
@@ -143,7 +143,7 @@ class minuit_optimizer(OptimizerMixin):
         if minimizer.valid:
             # Extra call to hesse() after migrad() is always needed for good error estimates. If you pass a user-provided gradient to MINUIT, convergence is faster.
             minimizer.hesse()
-            hess_inv = minimizer.np_covariance()
+            hess_inv = minimizer.covariance()
 
         unc = None
         if return_uncertainties:

--- a/src/pyhf/optimize/opt_minuit.py
+++ b/src/pyhf/optimize/opt_minuit.py
@@ -58,17 +58,24 @@ class minuit_optimizer(OptimizerMixin):
             wrapped_objective = objective_and_grad
             jac = None
 
-        kwargs = dict(
-            fcn=wrapped_objective,
-            grad=jac,
-            start=init_pars,
-            error=step_sizes,
-            limit=init_bounds,
-            fix=fixed_bools,
-            print_level=self.verbose,
-            errordef=self.errordef,
-        )
-        return iminuit.Minuit.from_array_func(**kwargs)
+        # kwargs = dict(
+        #     fcn=wrapped_objective,
+        #     grad=jac,
+        #     start=init_pars,
+        #     error=step_sizes,
+        #     limit=init_bounds,
+        #     fix=fixed_bools,
+        #     print_level=self.verbose,
+        #     errordef=self.errordef,
+        # )
+        minuit = iminuit.Minuit(fcn=wrapped_objective, grad=jac)
+        minuit.start = init_pars
+        minuit.error = step_sizes
+        minuit.limit = init_bounds
+        minuit.fix = fixed_bools
+        minuit.print_level = self.verbose
+        minuit.errordef = self.errordef
+        return minuit
 
     def _minimize(
         self,

--- a/src/pyhf/optimize/opt_minuit.py
+++ b/src/pyhf/optimize/opt_minuit.py
@@ -147,7 +147,7 @@ class minuit_optimizer(OptimizerMixin):
 
         unc = None
         if return_uncertainties:
-            unc = minimizer.np_errors()
+            unc = minimizer.errors
 
         return scipy.optimize.OptimizeResult(
             x=minimizer.values,

--- a/tests/test_optim.py
+++ b/tests/test_optim.py
@@ -496,22 +496,34 @@ def test_init_pars_sync_fixed_values_scipy(mocker):
 def test_init_pars_sync_fixed_values_minuit(mocker):
     opt = pyhf.optimize.minuit_optimizer()
 
-    # patch all we need
-    from pyhf.optimize import opt_minuit
+    def f(x, y, z):
+        # Need a function of three parameters for below
+        return (x - 2) ** 2 + (y - 3) ** 2 + (z - 4) ** 2
 
-    minimizer = mocker.patch.object(opt_minuit, 'iminuit')
-    opt._get_minimizer(None, [9, 9, 9], [(0, 10)] * 3, fixed_vals=[(0, 1)])
-    assert minimizer.Minuit.from_array_func.call_args[1]['start'] == [1, 9, 9]
-    assert minimizer.Minuit.from_array_func.call_args[1]['fix'] == [True, False, False]
+    # @kratsg, What is the need for mocker here?
+    # # patch all we need
+    # from pyhf.optimize import opt_minuit
+    #
+    # minimizer = mocker.patch.object(opt_minuit, 'iminuit')
+    minuit = opt._get_minimizer(f, [9, 9, 9], [(0, 10)] * 3, fixed_vals=[(0, 1)])
+    # assert minimizer.Minuit.from_array_func.call_args[1]['start'] == [1, 9, 9]
+    # assert minimizer.Minuit.from_array_func.call_args[1]['fix'] == [True, False, False]
+    assert list(minuit.values) == [1, 9, 9]
+    assert list(minuit.fixed) == [True, False, False]
 
 
 def test_step_sizes_fixed_parameters_minuit(mocker):
     opt = pyhf.optimize.minuit_optimizer()
 
-    # patch all we need
-    from pyhf.optimize import opt_minuit
+    def f(x, y, z):
+        # Need a function of three parameters for below
+        return (x - 2) ** 2 + (y - 3) ** 2 + (z - 4) ** 2
 
-    minimizer = mocker.patch.object(opt_minuit, 'iminuit')
-    opt._get_minimizer(None, [9, 9, 9], [(0, 10)] * 3, fixed_vals=[(0, 1)])
-    assert minimizer.Minuit.from_array_func.call_args[1]['fix'] == [True, False, False]
-    assert minimizer.Minuit.from_array_func.call_args[1]['error'] == [0.0, 0.01, 0.01]
+    # @kratsg, What is the need for mocker here?
+    # # patch all we need
+    # from pyhf.optimize import opt_minuit
+    #
+    # minimizer = mocker.patch.object(opt_minuit, 'iminuit')
+    minuit = opt._get_minimizer(f, [9, 9, 9], [(0, 10)] * 3, fixed_vals=[(0, 1)])
+    assert list(minuit.fixed) == [True, False, False]
+    assert list(minuit.errors) == [0.0, 0.01, 0.01]

--- a/tests/test_optim.py
+++ b/tests/test_optim.py
@@ -501,9 +501,10 @@ def test_init_pars_sync_fixed_values_minuit(mocker):
 
     minuit = mocker.patch.object(getattr(opt_minuit, 'iminuit'), 'Minuit')
     minimizer = opt._get_minimizer(None, [9, 9, 9], [(0, 10)] * 3, fixed_vals=[(0, 1)])
-
     assert minuit.called
-    assert minuit.call_args.args[1] == [1, 9, 9]
+    # python 3.6 does not have ::args attribute on ::call_args
+    # assert minuit.call_args.args[1] == [1, 9, 9]
+    assert minuit.call_args[0][1] == [1, 9, 9]
     assert minimizer.fixed == [True, False, False]
 
 

--- a/tests/test_optim.py
+++ b/tests/test_optim.py
@@ -496,34 +496,26 @@ def test_init_pars_sync_fixed_values_scipy(mocker):
 def test_init_pars_sync_fixed_values_minuit(mocker):
     opt = pyhf.optimize.minuit_optimizer()
 
-    def f(x, y, z):
-        # Need a function of three parameters for below
-        return (x - 2) ** 2 + (y - 3) ** 2 + (z - 4) ** 2
+    # patch all we need
+    from pyhf.optimize import opt_minuit
 
-    # @kratsg, What is the need for mocker here?
-    # # patch all we need
-    # from pyhf.optimize import opt_minuit
-    #
-    # minimizer = mocker.patch.object(opt_minuit, 'iminuit')
-    minuit = opt._get_minimizer(f, [9, 9, 9], [(0, 10)] * 3, fixed_vals=[(0, 1)])
-    # assert minimizer.Minuit.from_array_func.call_args[1]['start'] == [1, 9, 9]
-    # assert minimizer.Minuit.from_array_func.call_args[1]['fix'] == [True, False, False]
-    assert list(minuit.values) == [1, 9, 9]
-    assert list(minuit.fixed) == [True, False, False]
+    minuit = mocker.patch.object(getattr(opt_minuit, 'iminuit'), 'Minuit')
+    minimizer = opt._get_minimizer(None, [9, 9, 9], [(0, 10)] * 3, fixed_vals=[(0, 1)])
+
+    assert minuit.called
+    assert minuit.call_args.args[1] == [1, 9, 9]
+    assert minimizer.fixed == [True, False, False]
 
 
 def test_step_sizes_fixed_parameters_minuit(mocker):
     opt = pyhf.optimize.minuit_optimizer()
 
-    def f(x, y, z):
-        # Need a function of three parameters for below
-        return (x - 2) ** 2 + (y - 3) ** 2 + (z - 4) ** 2
+    # patch all we need
+    from pyhf.optimize import opt_minuit
 
-    # @kratsg, What is the need for mocker here?
-    # # patch all we need
-    # from pyhf.optimize import opt_minuit
-    #
-    # minimizer = mocker.patch.object(opt_minuit, 'iminuit')
-    minuit = opt._get_minimizer(f, [9, 9, 9], [(0, 10)] * 3, fixed_vals=[(0, 1)])
-    assert list(minuit.fixed) == [True, False, False]
-    assert list(minuit.errors) == [0.0, 0.01, 0.01]
+    minuit = mocker.patch.object(getattr(opt_minuit, 'iminuit'), 'Minuit')
+    minimizer = opt._get_minimizer(None, [9, 9, 9], [(0, 10)] * 3, fixed_vals=[(0, 1)])
+
+    assert minuit.called
+    assert minimizer.fixed == [True, False, False]
+    assert minimizer.errors == [0.0, 0.01, 0.01]

